rfcs/0000-rfc-calling-context-attribute.md

```diff
@@ -0,0 +1,116 @@
+      bar()
+    }
+}
+```
+
+The above code example shows a _calling context attribute_ annotating a function item, `fn bar()`, restricting it to be only called in contract contexts.
+
+Generally, Sway programmers should use this to annotate functions that are only applicable when being called from a contract, so the compiler can provide an error when they are misused.
+
+This annotation is transitive, meaning it needs to be also specified through all functions
+that call `contract_only` functions.
+
+
+# Reference-level explanation
+
+[reference-level-explanation]: #reference-level-explanation
+
+Right now the current prototype implementation for this feature is using the same approach that was implemented for storage access attributes.
+
+The attributes are parsed, and during type checking, we keep track of the current calling context for a particular type check context. Then during function and method applications, we can check if a particular function is missing a context attribute, and issue an error if that is the case.
+
+Later, as we do for purity, once the entire program is type-checked, we perform an analysis to figure out which top-level declarations are being called from the wrong context and issue the corresponding compiler error.
+
+The feature is pretty self-contained and doesn't affect any pre-existing code.
+
+The only expected potential breaking changes are once the standard library code will take advantage of this, then user code will need to be annotated further up the call chain.
+
+One thing to note is that is that functions that call an `contract_only` function needs themselves to be annotated as contract-only.
+
+So for instance imagine the following library code:
+
+```rust
+#[calling_context(contract_only)]
+fn foo() {}
+
+fn bar() {
+  foo();
+}
+```
+
+This will give a compile error, forcing the user to annotate `bar` with `#[calling_context(contract_only)]` as well.
+
+The only exception to this are contract-level functions which are themselves already satisfy the contract-only contraints by default.
+
+# Drawbacks
+
+[drawbacks]: #drawbacks
+
+The only drawback that comes to mind is the additional complexity to the language.
+
+But the extra safety guarantees provided by the feature should make up for it.
+
+# Rationale and alternatives
+
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+We could go with an `contract_only` keyword I guess, but re-using attributes seems like a better approach since we are already using the same scheme for storage access.
+
+# Prior art
+
+[prior-art]: #prior-art
+
+The obvious instance is Sway's very own storage attributes, which work used in a very similar manner, and from which the prototype implementation approach of this feature in the compiler was derived.
+
+# Unresolved questions
+
+[unresolved-questions]: #unresolved-questions
+
+1. Is `contract_only` the only calling context we want to support or can you think of some other we may want to support in the future?
```

Comment:
No, we also need predicate-only and script+contract-only (i.e. not-predicate).

- Predicate-only would be something like "get currently verifying predicate."
- Contract-only would be "read/write storage slot."
- Script+contract-only would be something like "contract call."